src/openrct2-ui/windows/TrackDesignPlace.cpp

```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
-        curTrackRotation &= 3;
-        if (track_coordinate->rotation_end & 4)
-        {
-            curTrackRotation |= 4;
-        }
-        if (!(curTrackRotation & 4))
-        {
-            curTrackStart += CoordsDirectionDelta[curTrackRotation];
-        }
+    bool DrawMiniPreviewIsPixelInBounds(const ScreenCoordsXY& pixel)
+    {
+        return pixel.x >= 0 && pixel.y >= 0 && pixel.x <= 160 && pixel.y <= 75;
     }
 
-    // Draw entrance and exit preview.
-    for (const auto& entrance : td6->entrance_elements)
+    uint8_t* DrawMiniPreviewGetPixelPtr(const ScreenCoordsXY& pixel)
     {
-        auto rotatedAndOffsetEntrance = origin + CoordsXY{ entrance.x, entrance.y }.Rotate(rotation);
+        return &_window_track_place_mini_preview[pixel.y * TRACK_MINI_PREVIEW_WIDTH + pixel.x];
+    }
 
-        if (pass == 0)
-        {
-            min.x = std::min(min.x, rotatedAndOffsetEntrance.x);
-            max.x = std::max(max.x, rotatedAndOffsetEntrance.x);
-            min.y = std::min(min.y, rotatedAndOffsetEntrance.y);
-            max.y = std::max(max.y, rotatedAndOffsetEntrance.y);
-        }
-        else
+    GameActions::Result FindValidTrackDesignPlaceHeight(CoordsXYZ& loc, uint32_t newFlags)
+    {
+        GameActions::Result res;
+        for (int32_t i = 0; i < 7; i++, loc.z += 8)
         {
-            auto pixelPosition = DrawMiniPreviewGetPixelPosition(rotatedAndOffsetEntrance);
-            if (DrawMiniPreviewIsPixelInBounds(pixelPosition))
+            auto tdAction = TrackDesignAction(CoordsXYZD{ loc.x, loc.y, loc.z, _currentTrackPieceDirection }, *_trackDesign);
+            tdAction.SetFlags(newFlags);
+            res = GameActions::Query(&tdAction);
+
+            // If successful don't keep trying.
+            // If failure due to no money then increasing height only makes problem worse
+            if (res.Error == GameActions::Status::Ok || res.Error == GameActions::Status::InsufficientFunds)
             {
-                uint8_t* pixel = DrawMiniPreviewGetPixelPtr(pixelPosition);
-                uint8_t colour = entrance.isExit ? _PaletteIndexColourExit : _PaletteIndexColourEntrance;
-                for (int32_t i = 0; i < 4; i++)
-                {
-                    pixel[338 + i] = colour; // x + 2, y + 2
-                    pixel[168 + i] = colour; //        y + 1
-                    pixel[2 + i] = colour;   // x + 2
-                    pixel[172 + i] = colour; // x + 4, y + 1
-                }
+                return res;
             }
         }
+        return res;
     }
-}
+};
 
-static void WindowTrackPlaceDrawMiniPreviewMaze(
-    TrackDesign* td6, int32_t pass, const CoordsXY& origin, CoordsXY min, CoordsXY max)
+rct_window* WindowTrackPlaceOpen(const track_design_file_ref* tdFileRef)
 {
-    uint8_t rotation = (_currentTrackPieceDirection + get_current_rotation()) & 3;
-    for (const auto& mazeElement : td6->maze_elements)
+    _trackDesign = TrackDesignImport(tdFileRef->path);
```

Comment:
the trackdesign should become owned by the window. I suggest you add a public method to the window called `Initialise(std::unique_ptr<TrackDesign>&& trackDesign)` and you move the loaded track design into the window.